cirq/optimizers/three_qubit_decomposition.py

```diff
@@ -0,0 +1,272 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Union
+
+import numpy as np
+
+import cirq
+from cirq import optimizers as opt
+from cirq import ops
+
+
+def three_qubit_matrix_to_operations(
+    q0: ops.Qid, q1: ops.Qid, q2: ops.Qid, u: np.ndarray, atol: float = 1e-8
+) -> List[ops.Operation]:
+    """Returns operations for a 3 qubit unitary.
+
+    The algorithm is described in Shende et al.:
+    Synthesis of Quantum Logic Circuits. Tech. rep. 2006,
+    https://arxiv.org/abs/quant-ph/0406176
+
+    Args:
+        q0: first qubit
+        q1: second qubit
+        q2: third qubit
+        u: unitary matrix
+        atol: A limit on the amount of absolute error introduced by the
+            construction.
+
+    Returns:
+        The resulting operations will have only known two-qubit and one-qubit
+        gates based operations, namely CZ, CNOT and rx, ry, PhasedXPow gates.
+    """
+    assert np.shape(u) == (8, 8)
+    assert cirq.is_unitary(u)
+
+    try:
+        from scipy.linalg import cossin
+    except ImportError:  # coverage: ignore
+        # coverage: ignore
+        raise ImportError(
+            "cirq.three_qubit_unitary_to_operations requires "
+            "SciPy 1.5.0+, as it uses the cossin function. Please"
+            " upgrade scipy in your environment to use this "
+            "function!"
+        )
```

Comment:
Because we do not want to impose for this single functionality a lower bound on the full project until we are convinced that 1.5.0 is propagated to the majority of large, slower moving environments (e.g. ours internally, or national labs). I can create an issue for this and make a TODO for a year from now or so. 1.5.0 is very new, it just came out in July.

Reply:
SGTM